runtime/gc_base/StringTable.cpp

```diff
@@ -976,4 +975,57 @@ j9gc_internString(J9VMThread *vmThread, j9object_t sourceString)
 	return internedString;
 }
 
+typedef struct {
+	J9UTF8* utf;
+	j9object_t stringObject;
+} J9UTFCacheEntry;
+
+static UDATA
+utfCacheHashFn(void *key, void *userData)
+{
+	/* UTF8 pointers are 2-aligned, so shift the pointer for maximum entropy */
+	return ((UDATA)((J9UTFCacheEntry*)key)->utf) >> 1;
+}
+
+static UDATA
+utfCacheHashEqualFn(void *leftKey, void *rightKey, void *userData)
+{
+	return ((J9UTFCacheEntry*)leftKey)->utf == ((J9UTFCacheEntry*)rightKey)->utf;
+}
+
+j9object_t
+j9gc_createJavaLangStringFromUTF(J9VMThread *vmThread, J9UTF8 *utf, UDATA stringFlags)
+{
+	bool cacheString = J9_ARE_ANY_BITS_SET(stringFlags, J9_STR_INTERN);
+	J9HashTable *utfCache = vmThread->utfCache;
+	j9object_t result = NULL;
+
+	if (cacheString) {
+		if (NULL == utfCache) {
+			J9JavaVM *vm = vmThread->javaVM;
+			utfCache = hashTableNew(OMRPORT_FROM_J9PORT(vm->portLibrary), J9_GET_CALLSITE(), 0, sizeof(J9UTFCacheEntry), sizeof(void*), 0, J9MEM_CATEGORY_VM, utfCacheHashFn, utfCacheHashEqualFn, NULL, NULL);
+			if (NULL == utfCache) {
+				vm->internalVMFunctions->setNativeOutOfMemoryError(vmThread, 0, 0);
+				goto done;
+			}
+			vmThread->utfCache = utfCache;
```

Comment:
Are we guaranteed to be single threaded when this is set?  Otherwise we may in the worst case leak a hashtable

Reply:
This code is now gone - please take another look. The cache applies to the current thread only, never a foreign one - it is only read or written on the current thread.